architecture/2024-08-bicepdeployment-controller.md

```diff
@@ -0,0 +1,446 @@
+
+#### Bran can use GitOps to deploy and manage a Radius application in production
+
+Bran is a developer who uses GitOps to automate the deployment of Radius applications to production Kubernetes clusters. His infrastructure team has set up a Radius installation and environment on the production cluster, and Bran wants to deploy his application using Radius. He has authored a Bicep manifest that defines the resources for his application and a Bicep parameters file that specifies the parameters for the manifest. He runs the `rad bicep generate bicepdeployment app.bicep --parameters app.bicepparam` command on his machine to create an BicepDeployment resource that contains the template and parameters specified. He then commits the BicepDeployment resource to the Git repository.
+
+When the GitOps tool watches the repository and applies the BicepDeployment resource to the production cluster, the Radius resources defined in the `app.bicep` Bicep manifest are deployed.
+
+#### Sam can use GitOps to patch a Radius application in production
+
+Sam is an SRE who is responsible for maintaining a Radius application in production. He notices that the application is experiencing performance issues and the number of application container replicas needs to be increased. He updates the Bicep parameters file for the application to specify the new number of replicas and runs the `rad bicep generate bicepdeployment app.bicep --parameters app.bicepparam` command on his machine to create an BicepDeployment resource that contains the updated parameters. He then commits the BicepDeployment resource to the Git repository.
+
+When the GitOps tool watches the repository and applies the BicepDeployment resource to the production cluster, the Radius resources defined in the `app.bicep` Bicep manifest are updated with the new parameters. The application container replicas are increased, and the performance issues are resolved.
+
+
+## User Experience
+<!--
+If the change impacts the user experience, provide expected interaction 
+flow we aim to achieve through this proposal.
+
+When users interact with Radius through the CLI, include sample 
+input commands and their corresponding output. Include a bicep/helm code 
+sample, if this proposal involves updates to that experience.
+-->
+
+**Sample Input:**
+<!--
+Provide a sample CLI command input and/or bicep/helm code.
+-->
+
+#### `env.bicep`
+```bicep
+extension radius
+
+param k8sNamespace string
+
+resource env 'Applications.Core/environments@2023-10-01-preview' = {
+  name: 'env'
+  properties: {
+    compute: {
+      kind: 'kubernetes'
+      namespace: k8sNamespace
+    }
+  }
+}
+```
+
+#### `env.bicepparam`
+```bicep
+using './env.bicep'
+
+param k8sNamespace string = 'demo'
+```
+
+**Sample Output:**
+<!--
+Provide a sample output for the inputs provided above.
+-->
+```
+> rad bicep generate bicepdeployment env.bicep --parameters env.bicepparam --out env.yaml
+
+Generating BicepDeployment resource...
```

Comment:
If "CRD" is a meaningful term for a user, and this command is generating a CRD, should we include that term in the command output and documentation?

Reply:
agreed